protocol/block.go

```diff
@@ -63,58 +69,86 @@ func (c *Chain) GenerateBlock(ctx context.Context, prev *bc.Block, snapshot *sta
 		},
 	}
 
+	var txEntries []*bc.TxEntries
+
 	for _, tx := range txs {
 		if len(b.Transactions) >= maxBlockTxs {
 			break
 		}
 
-		// TODO(jackson): Should this go in ConfirmTx too?
-		err = c.checkIssuanceWindow(tx)
+		// Filter out transactions that are not well-formed.
+		err := c.ValidateTx(tx.TxEntries)
 		if err != nil {
+			// TODO(bobg): log this?
 			continue
 		}
 
-		if validation.ConfirmTx(result, c.InitialBlockHash, bc.NewBlockVersion, timestampMS, tx) == nil {
-			err = validation.ApplyTx(result, tx)
-			if err != nil {
-				return nil, nil, err
-			}
-			b.Transactions = append(b.Transactions, tx)
+		// Filter out double-spends etc.
+		err = newSnapshot.ApplyTx(tx.TxEntries)
+		if err != nil {
+			// TODO(bobg): log this?
+			continue
 		}
+
+		b.Transactions = append(b.Transactions, tx)
+		txEntries = append(txEntries, tx.TxEntries)
 	}
-	b.TransactionsMerkleRoot, err = bc.MerkleRoot(b.Transactions)
+
+	var err error
+
+	b.TransactionsMerkleRoot, err = bc.MerkleRoot(txEntries)
 	if err != nil {
 		return nil, nil, errors.Wrap(err, "calculating tx merkle root")
 	}
-	b.AssetsMerkleRoot = result.Tree.RootHash()
-	return b, result, nil
+
+	b.AssetsMerkleRoot = newSnapshot.Tree.RootHash()
+
+	return b, newSnapshot, nil
+}
+
+// ValidateBlock validates an incoming block in advance of applying it
+// to a snapshot (with ApplyValidBlock) and committing it to the
+// blockchain (with CommitAppliedBlock).
+func (c *Chain) ValidateBlock(block, prev *bc.Block) error {
+	return c.validateBlock(block, prev, true)
+}
+
+func (c *Chain) validateBlock(block, prev *bc.Block, runProg bool) error {
```

Comment:
I think @jbowens is right on about the general point. It's not necessarily accurate to characterize the alternative as "pulling out fields and passing those" since the caller might not have the composite object to begin with. It might be more work to gin up or plumb in that composite object vs just using the one or two relevant values.

But in this case it seems okay, since Chain.ValidateBlock is just a convenience wrapper around validation.ValidateBlock. There's potentially a valid debate around whether we ought to have that convenience wrapper at all… a more interesting question than how many places do we call validation.ValidateBlock is how many places (in non-test code) do we call Chain.ValidateBlock? If it's just one or two then I'd advocate for inlining the convenience wrapper.